### ISSUE TITLE
Family Attendance Badge (24 month chart) shows incorrect data when manually taking attendance for a group.

### ISSUE THREAD

**Issue by hillsideassembly** [SparkDevNetwork/Rock#3985]
### Description
When entering attendance manually for a child, the Family Attendance Badge for that child reports wildly incorrect data.  This happens when adding attendance manually for a check-in group or for a regular group.

Today is November 7, 2019, so there has only been one week in November so far this year.  I added one attendance entry for Noah Decker for this month, and two for the previous months.  When looking at the badge though, his attendance is--literally--off the chart for November.

![noahdecker-history](https://user-images.githubusercontent.com/6493158/68415014-fce67b80-0156-11ea-9d35-e747a1495dc2.png)

![noahdecker-profile](https://user-images.githubusercontent.com/6493158/68415032-0b349780-0157-11ea-8b9c-d5e807979dbb.png)

I tried the same thing with Alex Decker, but entered the attendance manually to the "Grades K-1"
check-in group.  The attendance isn't off the chart this time, but November is higher than October, even though November has 1 entry and October has 2.

![alexdecker-history](https://user-images.githubusercontent.com/6493158/68415119-420aad80-0157-11ea-9c0b-0f0dcb1a0051.png)

![alexdecker-profile](https://user-images.githubusercontent.com/6493158/68415125-459e3480-0157-11ea-9372-8477d367cae5.png)

### Steps to Reproduce
1. In Group Viewer, select a group. Go to group Attendance.
1. Click the plus sign to add a new attendance.
1. Make sure the block settings for the Group Attendance Detail block has "Allow Adding Person" set to "Yes".
1. Use the "Add New Attendee" to add a child to that attendance using a date in the current month.  Save, then add more attendance entries in a previous month, and even further back if you wish.
1. Look at the Family Attendance badge for that child and compare how it looks with the actual data you entered.

**Expected behavior:**

The bars in the graph should correspond with the data even if manually entered.

**Actual behavior:**

The bars are displaying differently than the data shows.

### Versions
* **Rock Version:** 9.3 on demo. 9.4 on my church's rock.
* **Client Culture Setting:** en-US

**Comment by nairdo** [comment]
@hillsideassembly I have not been able to reproduce this issue (even on the demo site or on an older v9.3 demo system).  

That badge uses an underlying stored procedure to build the data for the chart and I'm wondering what you see when you ask for Noah Decker (ID 6 on my test system) data for the past 5 months like this:

```Sql
DECLARE @RC int
DECLARE @PersonId int = 6
DECLARE @RoleGuid uniqueidentifier
DECLARE @ReferenceDate datetime
DECLARE @MonthCount int = 5

EXECUTE @RC = [dbo].[spCheckin_BadgeAttendance] 
   @PersonId
  ,@RoleGuid
  ,@ReferenceDate
  ,@MonthCount
```

![image](https://user-images.githubusercontent.com/383807/71695045-34942d00-2d6e-11ea-87b7-a8c9a5039bad.png)

**Comment by hillsideassembly** [comment]
I logged into the demo site today and noticed it is still occurring before I even did anything.  
https://rock.rocksolidchurchdemo.com/Person/6/History

Someone renamed him to "Noah Smith" but it's the same person.
![2020-01-10 (2)](https://user-images.githubusercontent.com/6493158/72192802-8bcc7a00-33cc-11ea-8a24-82f70f73bc31.png)

I see under his attendance history that he has been unusually active at Rock Solid Church, so maybe that's part of the issue.
![noahsmith](https://user-images.githubusercontent.com/6493158/72193122-aeab5e00-33cd-11ea-84c2-4fdcc25fb144.png)

I ran your query and surprisingly it didn't show that many check ins for him though.
![query](https://user-images.githubusercontent.com/6493158/72193178-e61a0a80-33cd-11ea-8bbf-dec7c98be766.png)

I looked up a student in my church's rock (still running 9.4) and added a manual attendance for January.  You'll notice on his chart that it looks like he's been more active in January (this month) than in December...
![Annotation 2020-01-10 173542](https://user-images.githubusercontent.com/6493158/72193618-eb785480-33cf-11ea-888b-535ff778bdcb.png)

But that's not what his history says.  It shows he attended once in January, twice in December, and twice in November.  The chart should show the current month as smaller than the last two, and those two should be the same height.
![student](https://user-images.githubusercontent.com/6493158/72193668-1b275c80-33d0-11ea-84d7-0540c3c1f7f1.png)

Here is what his SQL query shows. I don't know why it says he attended 3 times in November, as you can see his history doesn't show that.  Either way, it shows that January is only 1, so again, it should be the smallest bar.  
![studentquery](https://user-images.githubusercontent.com/6493158/72193713-48740a80-33d0-11ea-94b6-c43521895be8.png)

Back to the demo site.  Maddie Lowe didn't have any attendance entered recently, so I added a couple, one for December, one for January.  Why aren't the bars the same height?
https://rock.rocksolidchurchdemo.com/Person/13/History
![maddiebars](https://user-images.githubusercontent.com/6493158/72194041-ba008880-33d1-11ea-8099-3b5a80f6ff38.png)
![maddiehistory](https://user-images.githubusercontent.com/6493158/72194065-d4d2fd00-33d1-11ea-8599-6a5c1ddebf2b.png)
![maddiesqlquery](https://user-images.githubusercontent.com/6493158/72194100-f46a2580-33d1-11ea-8469-02e1acd348bd.png)

I'm not sure why you weren't able to reproduce these on the demo site.  If you follow these steps, you should see that the bars on the badge aren't following the data correctly, especially for the current month, but also for the other months as well.

**Comment by JoshuaMaiorino** [comment]
I've seen a bug in this code before that doesn't filter out attendance records that don't have "DidAttend" flagged.  Not sure if that helps.

**Comment by nairdo** [comment]
@hillsideassembly We discovered bad data on the Rock Demo site that prevents an accurate attendance badge chart.  (_We don't have a easy way to fix that bad data yet since the demo site is restored from image each night, but we will address that bad data in the future._)

Regarding the issue you're describing on your 9.x system:
>  The chart should show the current month as smaller than the last two, and those two should be the same height.

The way the system currently works is that it will only report how many Sundays have occurred since the start of the month.  For example, if you're looking at September during the first week, it reports there being only 1 week (SundaysInMonth) in Sept as seen here:

![image](https://github.com/SparkDevNetwork/Rock/assets/383807/83d49c4b-9c5b-4bb0-bda5-c477e61b1129)

So, that should be kept in mind when considering this issue since the bar shows the number as a percentage of the total number of Sunday's in the month. 

That is currently by design (how it's always behaved).

However, there have been a few other bugs fixed since Rock v9.4 and we consider the original bug to be resolved.

If you encounter other bugs with the attendance badge bar in Rock v15.1 (or higher), please open another issue and provide data similar to what you've provided above.

Thanks for helping us make Rock better together!
